File: src/library/java/gg/generations/rarecandy/renderer/rendering/StateManager.java
package gg.generations.rarecandy.renderer.rendering;

import gg.generations.rarecandy.pokeutils.BlendType;
import gg.generations.rarecandy.pokeutils.CullType;
import org.lwjgl.opengl.GL11;

public class StateManager {
    private final StateToggle blend;
    private final StateToggle cull;
    private final StateToggle depth;

    public StateManager(Runnable blendEnable, Runnable blendDisable, Runnable cullEnable, Runnable cullDisable, Runnable depthEnable, Runnable depthDisable) {
        blend = new StateToggle(blendEnable, blendDisable);
        cull = new StateToggle(cullEnable, cullDisable);
        depth = new StateToggle(depthEnable, depthDisable);
    }

    public void toggle(boolean isBlend, boolean isCull, boolean isDepth) {
        blend.toggle(isBlend);
        cull.toggle(isCull);
        depth.toggle(isDepth);
    }

    public void toggle(RenderStage stage) {
        toggle(stage.isBlend(), stage.isCull(), stage.isDepthTest());
    }

    public void reset() {
        toggle(false, false, false);
    }
}

File: src/library/java/gg/generations/rarecandy/renderer/rendering/StateManager.kt
package gg.generations.rarecandy.renderer.rendering

class StateManager(
    blendEnable: () -> Unit,
    blendDisable: () -> Unit,
    cullEnable: () -> Unit,
    cullDisable: () -> Unit,
    depthEnable: () -> Unit,
    depthDisable: () -> Unit
) {
    private val blend: StateToggle = StateToggle(blendEnable, blendDisable)
    private val cull: StateToggle = StateToggle(cullEnable, cullDisable)
    private val depth: StateToggle = StateToggle(depthEnable, depthDisable)

    fun toggle(isBlend: Boolean, isCull: Boolean, isDepth: Boolean) {
        blend.toggle(isBlend)
        cull.toggle(isCull)
        depth.toggle(isDepth)
    }

    fun toggle(stage: RenderStage) = toggle(stage.isBlend, stage.isCull, stage.isDepthTest)

    fun reset() = toggle(isBlend = false, isCull = false, isDepth = false)
}
